alf/environments/wrappers.py

```diff
@@ -27,44 +30,161 @@
+                Dict env.observation_space.  If specified, only use the
+                spaces corresponding to the keys in FrameStack.  If is None,
+                stack all recognized spaces.
+                When input space is not a Dict, we just try to stack if we can.
+                Fields_to_stack doesn't apply anymore.
         """
         super().__init__(env)
-        self._frames = collections.deque(maxlen=stack_size)
+        self._frames = deque(maxlen=stack_size)
         self._channel_order = channel_order
         self._stack_size = stack_size
-
-        space = self.env.observation_space
-        assert channel_order in ['channels_last', 'channels_first']
-
-        if channel_order == 'channels_last':
-            shape = list(space.shape[0:-1]) + [stack_size * space.shape[-1]]
+        raw_space = self.env.observation_space
+
+        if not fields_to_stack and isinstance(raw_space, gym.spaces.Dict):
+            fields_to_stack = raw_space.spaces.keys()
+
+        self._fields_to_stack = fields_to_stack
+
+        def _stack_space(sp):
+            if isinstance(sp, gym.spaces.Box):
+                # Shape of stacked_space is determined by low.shape
+                low = np.concatenate([sp.low] * stack_size)
+                high = np.concatenate([sp.high] * stack_size)
+                assert channel_order in ['channels_last', 'channels_first']
+                if channel_order == 'channels_last':
+                    low = np.transpose(
+                        np.concatenate([np.transpose(sp.low)] * stack_size))
+                    high = np.transpose(
+                        np.concatenate([np.transpose(sp.high)] * stack_size))
+                stacked_space = gym.spaces.Box(
+                    low=np.array(low), high=np.array(high), dtype=sp.dtype)
+
+            elif isinstance(sp, gym.spaces.MultiDiscrete):
+                nvec = [sp.nvec] * stack_size
+                stacked_space = gym.spaces.MultiDiscrete(nvec)
+
+            else:
+                stacked_space = sp
+            return stacked_space
+
+        def _traverse(d, fields_to_stack=None, prefix=""):
+            assert isinstance(d, gym.spaces.Dict), 'input is not dict'
+            res = d
+            for name, sp in d.spaces.items():
+                if prefix:
+                    prefix += "."
+                path = prefix + name
+                transformed = sp
+                if isinstance(sp, gym.spaces.Dict):
+                    transformed, fields_to_stack = _traverse(
+                        sp, fields_to_stack, path)
+                else:
+                    if fields_to_stack is None or path in fields_to_stack:
+                        transformed = _stack_space(sp)
+                        if fields_to_stack:
+                            fields_to_stack = [
+                                item for item in fields_to_stack
+                                if item != path
+                            ]
```

Comment:
why need L105-L109?

Reply:
This is to compute the remaining fields that we haven't stacked, for verification purposes, so that all fields specified in fields_to_stack have to appear in input obs.